ynr/apps/moderation_queue/management/commands/moderation_queue_detect_faces_in_queued_images.py

```diff
@@ -19,27 +23,85 @@ def handle(self, **options):
         attributes = ["ALL"]
 
         any_failed = False
-        qs = QueuedImage.objects.filter(decision="undecided").exclude(
-            face_detection_tried=True
-        )
+        qs = QueuedImage.objects.all()
 
         for qi in qs:
 
             try:
                 detected = rekognition.detect_faces(
                     Image={"Bytes": qi.image.file.read()}, Attributes=attributes
                 )
+
                 self.set_x_y_from_response(qi, detected, options["verbosity"])
             except Exception as e:
                 msg = "Skipping QueuedImage{id}: {error}"
                 self.stdout.write(msg.format(id=qi.id, error=e))
                 any_failed = True
 
             qi.face_detection_tried = True
+            self.detect_rotation(image=qi.image, detected=detected)
             qi.save()
         if any_failed:
             raise CommandError("Broken images found (see above)")
 
+    def detect_rotation(self, image, detected):
+        """
+        Detects the rotation of an image and returns an integer
+        """
+
+        # opens the image as a PIL.PngImagePlugin.PngImageFile
+        image = Image.open(image, formats=None)
+        # this step won't likely be used since we may be
+        # removing exif data in the png conversion step
+        try:
+            for orientation in ExifTags.TAGS.keys():
+                if ExifTags.TAGS[orientation] == "Orientation":
+                    break
+            exif = dict(image._getexif())
+
+            if exif[orientation] == 1:
+                print(270)
+                rotated = image.rotate(angle=270)
+                rotated.show()
+            else:
+                rotated = ImageOps.exif_transpose(image)
+        except:
+            # if roll is positive, then the image needs to be rotated 90 degrees clockwise
+            if detected["FaceDetails"][0]["Pose"]["Roll"] > 0:
+                print("Rotate 90 degrees clockwise")
+                rotated = image.rotate(angle=90)
+                rotated.show()
+            # if roll is negative, then the image needs to be rotated 90 degrees counter clockwise
+            elif detected["FaceDetails"][0]["Pose"]["Roll"] < 0:
+                print("Rotate 90 degrees counter clockwise")
+                rotated = image.rotate(angle=270)
+                rotated.show()
+            # if roll is 180, then the image is upside down
+            elif detected["FaceDetails"][0]["Pose"]["Roll"] == 180:
+                print("Upside down")
+                rotated = image.rotate(angle=180)
+                image.show()
+                # if roll is 0, then the image is upright
+            else:
+                print("Roll is:", detected["FaceDetails"][0]["Pose"]["Roll"])
+                print("No rotation needed")
+                pass
+
```

Comment:
I don't think this approach will do what you want. Splitting the images into categories `roll < 0`, `roll > 0` and `roll == 180` isn't going to work. For one thing, the `roll ==180` criteria will never be hit. There's also the issue that when `roll == -165` you probably still want to rotate the image by 180° rather than 165°.

Instead I think what you want to do is : "Calculate the angle image is rotated by rounding the roll value to the nearest multiple of 90". Some examples:
`-178 -> -180`
`75 -> 90`
`-12 -> 0`
etc. 
You can do this as: `img_rotation_angle = 90 * round(roll / 90)`

Reply:
Thanks! Hadn't gotten that far yet, but this is useful.

Reply:
Resolved with https://github.com/DemocracyClub/yournextrepresentative/pull/2015/commits/25833f96bf9ce2e2cef157b933bb7e47f9139100